packages/grafana-data/src/types/events.ts

```diff
@@ -0,0 +1,177 @@
+import { IHttpResponse } from 'angular';
+import { TimeRange } from './time';
+import { DashboardModel } from 'app/features/dashboard/state';
+import { DataFrame } from './dataFrame';
+
+export type AlertPayload = [string, string?];
+
+export interface ShowDashSearchPayload {
+  query?: string;
+}
+
+export interface LocationChangePayload {
+  href: string;
+}
+
+export interface ShowModalPayload {
+  model?: any;
+  modalClass?: string;
+  src?: string;
+  templateHtml?: string;
+  backdrop?: any;
+  scope?: any;
+}
+
+export interface ShowConfirmModalPayload {
+  title?: string;
+  text?: string;
+  text2?: string;
+  text2htmlBind?: boolean;
+  confirmText?: string;
+  altActionText?: string;
+  yesText?: string;
+  noText?: string;
+  icon?: string;
+
+  onConfirm?: () => void;
+  onAltAction?: () => void;
+}
+
+type DataSourceResponsePayload = IHttpResponse<any>;
+
+export interface SaveDashboardPayload {
+  overwrite?: boolean;
+  folderId?: number;
+  makeEditable?: boolean;
+}
+
+export interface PanelChangeViewPayload {
+  fullscreen?: boolean;
+  edit?: boolean;
+  panelId?: number;
+  toggle?: boolean;
+}
+
+export type RemovePanelPayload = number;
+
+export interface GraphHoverPayload {
+  pos: any;
+  panel: {
+    id: number;
+  };
+}
+
+export interface ToggleKioskModePayload {
+  exit?: boolean;
+}
+
+export interface GraphClickedPayload {
+  pos: any;
+  panel: any;
+  item: any;
+}
+
+export interface ThresholdChangedPayload {
+  threshold: any;
+  handleIndex: any;
+}
+
+export interface DashScrollPayload {
+  restore?: boolean;
+  animate?: boolean;
+  pos?: number;
+}
+
+export interface MenuElement {
+  text: string;
+  click: string;
+  role?: string;
+  shortcut?: string;
+}
+
+export interface AppEvent<T> {
+  readonly name: string;
+  payload?: T;
+}
+
+export function eventFactory<T>(name: string): AppEvent<T> {
```

Comment:
One thing that would probably make this a bit more safe is to check whether the names are unique. I think we already do that in redux stores with actions which can also clash.

Reply:
That's a good point! I'll implement that